### PR TITLE
feat(speculative): add Kimi K3 EAGLE-3 training

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds EAGLE-3 draft training for a frozen **Kimi K3** target, the EAGLE counterpart to the DSpark support in #3263.

Re-opens #3264, which GitHub closed automatically when its base branch `huiyingl/feat/kimi-k3` was merged and deleted. Same work, now re-targeted at `main`; the review history lives on #3264.

Kimi K3 interleaves KDA layers with **NoPE MLA** full-attention layers, so neither the dense EAGLE-3 draft nor the DeepSeek MLA draft (which always rotates the rope slice) can represent it:

| | DeepSeek MLA draft | Kimi K3 draft |
|---|---|---|
| rope slice | rotated (interleaved RoPE), per-TTT-step phase offset | carried unrotated, no rotary cache at all |
| attention output | `o_proj(attn)` | sigmoid output gate, then `o_proj` |
| MLP / norm | SwiGLU, shared RMSNorm helper | K3 `SituAndMul` MLP, K3 fp32-variance RMSNorm |

`KimiK3Eagle3DraftModel` mirrors exactly one K3 MLA layer (Q-LoRA, the compressed KV latent, the gate, the SiTU MLP); KDA, the 896 routed experts and the attention-residual mixer stay in the target. Everything EAGLE-3 specific (the `[embed, hidden]` fused first layer, the `cache_hidden` TTT recurrence with diagonal extension, `d2t`/`t2d`, `project_hidden_states`/`compute_logits`) is inherited from the DeepSeek MLA draft through one new `_build_inner_model` hook, so the trainer, the checkpointer and the serve-side contract are reused unchanged.

Positional information reaches the draft through the target hidden states it is conditioned on, exactly as it reaches the target's own MLA layers; `test_forward_is_position_id_invariant` pins that contract.

## Recipe wiring

K3 has no HuggingFace implementation and its experts do not fit on one GPU, so the colocated target is built from the (possibly nested) text config through AutoModel's custom expert-parallel path rather than the generic `from_pretrained`. A vision-language checkpoint is loaded as its text-only `KimiK3ForCausalLM`; the target config is a copy, so the draft config serialized into `config.json` is untouched.

Combinations K3 cannot serve are rejected up front, before the 93-layer target is loaded: non-colocated backends (SGLang/vLLM/remote have no K3 implementation), context and tensor parallelism (K3 owns CP end to end and declares `supports_tp=False`), pipeline parallelism (aux capture hooks one non-pipelined forward), sequence packing (K3 owns its packed-attention path), and P-EAGLE.

The target's expert dispatcher defaults to `torch`, the dispatcher K3's own SFT reference config runs expert parallelism with; `deepep`/`hybridep` stay opt-in through `target_dispatcher` since they need a matching DeepEP build.

## Also here

- `freeze_embeddings()` on the shared MLA draft base. The recipe calls it for every draft, and it was missing, so `freeze_embeddings: true` raised `AttributeError` on the DeepSeek draft too. It matters for K3: the draft copies a 163840 x 7168 embedding table.
- The packing error message now names the actual draft class instead of always saying `DeepseekV3Eagle3DraftModel`.

## Testing

- `tests/unit_tests/speculative/test_eagle3_kimi_k3_draft.py` (16 tests): forward/TTT/packing shapes, NoPE position invariance, the output gate, backbone composition, the GQA and `mla_use_nope` rejections, the vocab-remap and embedding-copy API, registry resolution for both architectures, and end-to-end training through the shared `Eagle3TrainerModule`.
- `tests/unit_tests/recipes/llm/test_eagle3_kimi_k3_target.py` (16 tests): backend defaults and overrides, every gate, and the resulting `from_config` call including the untouched draft base config.
- `tests/unit_tests/speculative/` and `tests/unit_tests/recipes/llm/` pass locally: the failure list is identical to the one `main` produces on its own (25 `test_benchmark.py` failures from a local environment gap, unrelated to this PR).
- No GPU run on my side: I have no Kimi K3 target weights, so everything above is CPU-verified only. As on #3263, asking @HuiyingLi to run the GPU validation; see the comment below for what a GPU run would cover that CPU cannot.

## Re-targeting onto main

`main` now carries the merged K3 model, so this branch merges `main` and keeps main's version of every K3 file. One fix was needed on top:

- The non-colocated backends call `_setup_online_target` with `draft_base_config=None`, so the K3 gate reading `.model_type` unconditionally raised `AttributeError` before the backend dispatch. It now reads through `getattr`. This is covered by main's `test_eagle3_vllm_backend.py` / `test_eagle3_sglang_backend.py`, which fail without it.

### CODE REVIEW

File: nemo_automodel/components/speculative/eagle/draft_kimi_k3.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        if self.use_output_gate:
+            self.g_proj = nn.Linear(in_features, self.num_heads * self.v_head_dim, bias=False)
+
+    def forward(
+        self,
+        combined_states: torch.Tensor,
+        attention_mask: torch.Tensor,
+        position_ids: torch.Tensor,
+        cache_hidden: list[list[torch.Tensor]],
+    ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/speculative/eagle/draft_kimi_k3.py:125]: Added Args/Returns to both forwards, documenting the fused [embed, hidden] input and the [K_list, V_list] TTT cache.

**Comment by khazic** [comment]
Merged the latest main to clear the conflict. The only conflict was in the import block of `train_eagle3.py`: main dropped `json` after moving the losses write to `save_losses`, while this branch had added `copy` and `json`. Kept `copy` (still used by the K3 target-config deepcopy) and dropped `json`. No other changes.

**Comment by HuiyingLi** [comment]
GPU validation completed successfully on the full Kimi K3 target.

### Run

- Validated implementation commit: `0514f0cb0f944592c32bfb16f8eacd0d699f3474`
- Recipe memory fix now on the PR branch: `e423803fa0148a6653adaa5a04f627b40f68d205`
- Slurm job: `6014892` — `COMPLETED (0:0)`, 52/52 training steps
- Topology: 64 GB200 GPUs (16 nodes x 4), FSDP2, EP64 / PP1 / CP1 / TP1
- Runtime: 2K sequence length, micro-batch size 1, `ttt_steps=7`, compressed 32K draft vocab, frozen draft embeddings, HybridEP target dispatcher, native K3 MLA + FLA KDA target, and the PR's eager EAGLE-3 MLA draft
- Peak sampled GPU memory: **182,827 MiB (178.54 GiB)**, leaving about 6.49 GiB sampled headroom
- W&B: https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/7ehs2nyu
- Rank-0 run log: `/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_llm/users/huiyingl/k3/pr3286/logs/full_ep64_s2048_52steps_6014892_nvl72142-T01.out`

### Result

- Train loss: `11.8153` at step 1 -> `0.5804` at step 52
- Train accuracy: `0.0001` -> `0.8242`
- Simulated train tau: `1.0000` -> `6.0829`
- Validation loss: `11.1324` after epoch 1 -> `0.5763` after epoch 13
- Validation accuracy: `0.0703` -> `0.8281`
- Validation simulated tau: `1.1608` -> `6.1290`
- The preceding one-node / 4-layer EP4 smoke also completed 8/8 steps in job `6014767` before the full run.

Final log excerpt:

```text
epoch=12 step=52 train_loss=0.580449 train_acc=0.824219 train_tau_sim=6.0829 lr=1.000e-05
epoch=12 val_loss=0.576336 val_acc=0.828125 val_tau_sim=6.1290
Training complete: global_step=52
```

`reshard_after_forward: true` is required for this frozen target. The identical full Kimi K3 FSDP target path was already measured to OOM near the 184 GiB device limit with `false`: the target runs under `torch.no_grad()`, so no backward pass releases the all-gathered block parameters. This EAGLE-3 run still reached 178.54 GiB with `true`, leaving only about 6.49 GiB sampled headroom. I pushed the recipe change in `e423803fa`.

One data caveat: this run used a small, repeatedly sampled local 256-train / 32-validation Tulu3 slice, not PerfectBlend responses regenerated from Kimi K3 at temperature 0.7. It validates full-target loading, EP64 execution, aux hidden-state capture, draft forward/backward, metric aggregation, memory behavior, and short-run optimization movement; the high tau/accuracy from repeatedly fitting this small slice should not be treated as final serving-quality data.

**Review by HuiyingLi — APPROVED** [review]
